compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -2756,6 +2788,61 @@ BType getTypeForUnionTypeMembersAssignableToType(BUnionType unionType, BType tar
         }
     }
 
+    BType getTypeForMapTypeMembersAssignableToType(BMapType mapType, BType targetType) {
+        if (targetType.tag != TypeTags.MAP && targetType.tag != TypeTags.RECORD) {
+            return symTable.semanticError;
+        }
+        if (targetType.tag == TypeTags.MAP) {
+            BType intersectionConstraintType = getTypeIntersection(mapType.constraint,
+                    ((BMapType) targetType).constraint);
+            if (intersectionConstraintType != symTable.semanticError) {
+                return symTable.semanticError;
+            }
+            mapType.constraint = intersectionConstraintType;
+            return mapType;
+        }
+        return getTypeForRecordTypeMembersAssignableToType((BRecordType) targetType, mapType);
+    }
+
+    BType getTypeForRecordTypeMembersAssignableToType(BRecordType recordType, BType targetType) {
+        if (targetType.tag != TypeTags.MAP && targetType.tag != TypeTags.RECORD) {
+            return symTable.semanticError;
+        }
+        if (targetType.tag == TypeTags.MAP) {
+            BRecordTypeSymbol recordSymbol =
+                    Symbols.createRecordSymbol(0, names.fromString("$anonRecordIntersectionType$" + recordCount++),
+                            env.enclPkg.symbol.pkgID, null, env.scope.owner, recordType.tsymbol.pos, VIRTUAL);
+            LinkedHashMap<String, BField> fields = new LinkedHashMap<>();
+            for (String key : recordType.fields.keySet()) {
+                BType intersectionKeyType = getTypeIntersection(recordType.fields.get(key).type,
+                        ((BMapType) targetType).constraint);
+                if (intersectionKeyType == symTable.semanticError) {
+                    return symTable.semanticError;
+                }
+                BVarSymbol fieldSymbol = new BVarSymbol(0, names.fromString(key), env.enclPkg.symbol.pkgID,
+                        intersectionKeyType, recordSymbol, recordType.fields.get(key).pos, COMPILED_SOURCE);
+                BField field = new BField(names.fromString(key), recordType.fields.get(key).pos, fieldSymbol);
+                fields.put(key, field);
+            }
+            BRecordType intersectionRecordType = new BRecordType(recordSymbol);
+            recordSymbol.type = intersectionRecordType;
+            intersectionRecordType.fields = fields;
+
+            if (recordType.restFieldType != null) {
+                BType intersectionRestType = getTypeIntersection(recordType.restFieldType,
+                        ((BMapType) targetType).constraint);
+                if (intersectionRestType == symTable.semanticError) {
+                    intersectionRecordType.sealed = true;
+                }
+                intersectionRecordType.restFieldType = intersectionRestType;
+            } else {
+                intersectionRecordType.sealed = recordType.sealed;
+            }
+            return intersectionRecordType;
+        }
+        return symTable.semanticError;
```

Comment:
Aren't we implementing record-record intersection here? If so shouldn't we change the check in L2808 to just

`targetType.tag != TypeTags.MAP`

@dulvinw is probably implementing record-record intersection though.

Reply:
This is added with the help of @dulvinw 's implementation.